src/CheckoutSdk/JsonSerializer.cs

```diff
@@ -1,17 +1,75 @@
+using Checkout.Common;
 using Checkout.Instruments.Four.Get.Util;
 using Checkout.Instruments.Four.Update.Util;
 using Checkout.Workflows.Four.Actions.Response.Util;
 using Checkout.Workflows.Four.Conditions.Response.Util;
+using System;
+using System.Collections.Generic;
+
+#if NET5_0_OR_GREATER
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+#else 
 using Newtonsoft.Json;
 using Newtonsoft.Json.Converters;
 using Newtonsoft.Json.Serialization;
-using System;
-using System.Collections.Generic;
+#endif
 
 namespace Checkout
 {
     public class JsonSerializer : ISerializer
     {
+
+#if NET5_0_OR_GREATER
+
+        private readonly JsonSerializerOptions _serializerSettings;
+
+        public JsonSerializer(Action<JsonSerializerOptions> configureSettings = null)
+        {
+            _serializerSettings = CreateSerializerSettings(configureSettings);
+        }
+
+
+        public object Deserialize(string payload, Type objectType)
+        {
+            CheckoutUtils.ValidateParams("payload", payload);
+            return System.Text.Json.JsonSerializer.Deserialize(payload, objectType, _serializerSettings);
+        }
+
+        public IDictionary<string, object> Deserialize(string payload)
+        {
+            return System.Text.Json.JsonSerializer.Deserialize<Dictionary<string, object>>(payload, _serializerSettings);
+        }
+
+        public string Serialize(object payload)
+        {
+            CheckoutUtils.ValidateParams("payload", payload);
+            return System.Text.Json.JsonSerializer.Serialize(payload, payload.GetType(), _serializerSettings);
```

Comment:
(replying here so you can reply directly)

Hi @oferns. Just to be clear, the problem/incompatibility I found was on **serialization**, that only applies when serializing request classes (PaymentRequest for example). As you mentioned polymorphic **deserialization** is possible, the solution you presented for response classes is pretty much the Newtonsoft way to do it, and that's all right.

Certain tests use Four credentials, some use OAuth, but a big amount enforce "default" (pk/sk_test_uuid) key credentials - the credentials you have. You should be able to reproduce it your credentials. Try the first test on `Checkout.Payments.RequestPaymentsIntegrationTest`. Put a breakpoint on `JsonSerializer` line 48 `Serialize(object payload)`. Enforce the execution with .NET >5. Have a look on the object to be serialized vs. the result of the serialization. 

There are a few other issues that I found yesterday, like Unicode serialization (the + on the timestamps were being serialized as [ub002B](https://github.com/dotnet/runtime/issues/35281)), serialization should enforce snake_case instead of CamelCase (not out of the box) - these are minor issues, there are quick fixes for them. Polymorphic serialization is the real blocker here.

Please let me know how it worked for you and feel free to suggest solutions / alternatives to the current serialization approach. We'd be happy to improve it. If you agree, for the time being, I strongly suggest to break up this problem in 2. Lets first tackle the .NET targeting and then with some more time we can come up with a better solution for serialization/deserialization. 

Thanks!

> Hi Raul, firstly thanks for your time on this. It is much appreciated. Whilst polymorphic deserialization is not supported out of the box with System.Text.Json it is fairly straightforward to implement. MS provides an example here <https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-converters-how-to?pivots=dotnet-6-0#support-polymorphic-deserialization> and this is pretty much how I have tried to implement it in classes like PaymentResponseDestinationConverter.cs. The premise is you clone the reader, find the property that defines the type, switch on that and pass the original reader to an appropriately typed deserialize method, as seen on lines 50-55 of PaymentResponseDestinationConverter.cs. I think that is pretty much the approach that Newtonsoft take. Unfortunately you cannot pass the same JSON options to the new serializer and that might be the problem here. If I had the ability to run the integration tests I am sure I can fix that aspect pretty quickly. That said, I am happy to exclude those changes and resubmit a PR just around the targeting and I will do that today. If I could get creds to run the tests then I can resubmit a further update for STJ integration for your consideration . Thanks, Ollie
> […](#)
> On Mon, 14 Feb 2022 at 17:20, Raul Estrela ***@***.***> wrote: ***@***.**** requested changes on this pull request. Thanks for your contribution. Overall, really nice work! For the moment, I won't point any detail in particular but I found a few incompatibilities / issues with integration of System.Text.Json causing most of tests to fail in >.NET5 being the most concerning to me the lack of support to polymorphic serialization. In JsonSerializer we rely heavily on the ability to serialize dynamic subtypes in runtime. For example, the supertype AbstractRequestSource being used at PaymentRequest has different subtypes (RequestIdSource, RequestCardSource...) meaning that an user can create a PaymentRequest and set different subtypes as a source. Newtonsoft is able to infer these subtypes in runtime (probably with a performance handicap) and generates the expected json, for example: { amount: 10, source: { type: "card", number: "123", cvv: "345" } } System.Text.Json does not consider the subtype in particular, only the parent. So we get something like this: { amount: 10, source: { } } I was reading about this topic and it seems there's no solution generic enough in place that allows such functionality out of the box like Newtonsoft. It seems that was a decision taken during the design of the library to increase the performance and make it less susceptible in terms of security. [dotnet/runtime#29937](https://github.com/dotnet/runtime/issues/29937) <[dotnet/runtime#29937](https://github.com/dotnet/runtime/issues/29937)> [dotnet/runtime#30083](https://github.com/dotnet/runtime/issues/30083) <[dotnet/runtime#30083](https://github.com/dotnet/runtime/issues/30083)> Open issue: [dotnet/runtime#63747](https://github.com/dotnet/runtime/issues/63747) <[dotnet/runtime#63747](https://github.com/dotnet/runtime/issues/63747)> Please let me know what do you think. If there's no quick solution for this issue we might consider not to adopt System.Text.Json by now, proceed only with the retargeting. Would that solution work for you? In case it sounds good for you please revert all the changes related to serialization and squash all the commits into 1 (git squash). Thanks! — Reply to this email directly, view it on GitHub <[#163 (review)](https://github.com/checkout/checkout-sdk-net/pull/163#pullrequestreview-881955323)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AARGOHGJRIEHLJIYABCU46LU3E2U5ANCNFSM5OER7BDA> . Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. You are receiving this because you authored the thread.Message ID: ***@***.***>